### PR TITLE
Simple merkle rfc compatibility

### LINKED ISSUE #1892: crypto/merkle: SimpleMerkle should differentiate between interior & leaf nodes
See 
https://github.com/tendermint/tmlibs/pull/211#issuecomment-391102658
https://github.com/tendermint/tmlibs/pull/211#issuecomment-391044446

Moved over from https://github.com/tendermint/go-crypto/issues/107 (contains more pointers).

### PR DESCRIPTION
**Pull request by ValarDragon**
Closes #1892 

What remains:
* [x] Fix current error with indexer in the lite client
* [x] Update docs
* [x] Add Trillian test vectors
* [x] Change the split point per RFC 6962 for efficiency of leaf additions

* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: crypto/merkle/rfc6962_test.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	leaf1, leaf2 := []byte("Hello"), []byte("World")
+	_, leafHashTrail := trailsFromByteSlices([][]byte{leaf1})
+	hash1 := leafHashTrail.Hash
+	_, leafHashTrail = trailsFromByteSlices([][]byte{leaf2})
+	hash2 := leafHashTrail.Hash
+	if bytes.Equal(hash1, hash2) {
+		t.Errorf("Leaf hashes should differ, but both are %x", hash1)
+	}
+	// Compute an intermediate subtree hash.
+	_, subHash1Trail := trailsFromByteSlices([][]byte{hash1, hash2})
```
Comment (Liamsi) [inline crypto/merkle/rfc6962_test.go]: I wonder if shouldn't also split up `trailsFromByteSlices` into `hashChildren` and `hashLeaf` or so. Orthogonal to this PR though.

File: crypto/merkle/simple_tree_test.go [hunk 2]
```diff
@@ -43,30 +42,53 @@ func TestSimpleProof(t *testing.T) {
 		require.Equal(t, proof.Total, total, "Unmatched totals: %d vs %d", proof.Total, total)
 
 		// Verify success
-		err := proof.Verify(rootHash, itemHash)
-		require.NoError(t, err, "Verificatior failed: %v.", err)
+		err := proof.Verify(rootHash, item)
+		require.NoError(t, err, "Verificatiom failed: %v.", err)
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply crypto/merkle/simple_tree_test.go]: LOL one typo changed to another typo

File: crypto/merkle/rfc6962_test.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	_, leafHashTrail := trailsFromByteSlices([][]byte{[]byte("L123456")})
+	leafHash := leafHashTrail.Hash
+	_, leafHashTrail = trailsFromByteSlices([][]byte{[]byte{}})
+	emptyLeafHash := leafHashTrail.Hash
+	for _, tc := range []struct {
+		desc string
+		got  []byte
+		want string
+	}{
+		// Check that the empty hash is not the same as the hash of an empty leaf.
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply crypto/merkle/rfc6962_test.go]: Not sure I follow, how would that ever happen. Thats also not a trillian test vector

File: crypto/merkle/rfc6962_test.go (hunk 1 above)
Comment (ValarDragon) [reply crypto/merkle/rfc6962_test.go]: Sounds like a good idea to me, but lets open up in a separate issue

File: crypto/merkle/rfc6962_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Liamsi) [reply crypto/merkle/rfc6962_test.go]: Sorry, I meant the first test-vector (empty root, not empty inner leaf; which indeed doesn't make any sense :-D)

File: crypto/merkle/hash.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package merkle
+
+import (
+	"github.com/tendermint/tendermint/crypto/tmhash"
+)
+
+// TODO: make these have a large predefined capacity
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply crypto/merkle/hash.go]: Made one! #2833

File: types/tx.go [hunk 5]
```diff
@@ -69,7 +70,7 @@ func (txs Txs) Proof(i int) TxProof {
 	l := len(txs)
 	bzs := make([][]byte, l)
 	for i := 0; i < l; i++ {
-		bzs[i] = txs[i]
+		bzs[i] = txs[i].Hash()
```
Comment (ebuchman) [inline types/tx.go]: Think we need to update this in the spec. It seems txs are treated differently from everything else by being hashed before passed to the MerkleTree. Should the other data structures be treated similarly, so they're hashed before being passed in (and eg. concatenated with 0) ?

File: types/tx.go [hunk 6]
```diff
@@ -87,8 +88,8 @@ type TxProof struct {
 	Proof    merkle.SimpleProof
 }
 
-// LeadHash returns the hash of the this proof refers to.
-func (tp TxProof) LeafHash() []byte {
+// Leaf returns the hash(tx), which is the leaf in the merkle tree which this proof refers to.
+func (tp TxProof) Leaf() []byte {
```
Comment (ebuchman) [inline types/tx.go]: We should track these in the GoAPI breaking changes

File: crypto/merkle/simple_proof.go [hunk 7]
```diff
@@ -67,7 +66,8 @@ func SimpleProofsFromMap(m map[string][]byte) (rootHash []byte, proofs map[strin
 
 // Verify that the SimpleProof proves the root hash.
 // Check sp.Index/sp.Total manually if needed
-func (sp *SimpleProof) Verify(rootHash []byte, leafHash []byte) error {
+func (sp *SimpleProof) Verify(rootHash []byte, leaf []byte) error {
```
Comment (ebuchman) [inline crypto/merkle/simple_proof.go]: Did we change the semantics of Verify here? It used to take the hash of the thing passed to SimpleMerkle, and now it takes the pre-image bytes?

File: crypto/merkle/simple_tree_test.go [hunk 8]
```diff
@@ -43,30 +42,53 @@ func TestSimpleProof(t *testing.T) {
...
 		require.Error(t, err, "Expected verification to fail for mutated leaf hash")
 
 		// Mutating the rootHash should make it fail.
-		err = proof.Verify(MutateByteSlice(rootHash), itemHash)
+		err = proof.Verify(MutateByteSlice(rootHash), item)
 		require.Error(t, err, "Expected verification to fail for mutated root hash")
 	}
 }
+
+func Test_getSplitPoint(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply crypto/merkle/simple_tree_test.go]: I use https://github.com/cweill/gotests to automate alot of the boiler plate of table driven tests. I forgot to change the template to remove the underscore though!

File: crypto/merkle/simple_proof.go (hunk 7 above)
Comment (ValarDragon) [reply crypto/merkle/simple_proof.go]: Yeah, we had to so that everything that uses it doesn't have to be aware of the 0 byte prefix.

**Comment by ebuchman** [comment]
@ValarDragon will this require updates on the SDK? Let's make sure those can get in if we're going to merge this for v0.27

**Comment by ValarDragon** [comment]
It would break the store package test cases, but other than that I don't think it breaks anything. I'm not sure if anyone will have time to update the SDK though.

**Comment by ebuchman** [comment]
> It would break the store package test cases, but other than that I don't think it breaks anything. I'm not sure if anyone will have time to update the SDK though.

Ok - I don't think we can bring it in if we're not going to take the time to fix the tests. It might also break lite client code if Verify is expecting something different. So I think this is justification to hold off for v0.28. We should look to include all final breaking changes for release in v0.28 if possible, ie. this, #1007 , and possibly #2671

**Comment by ebuchman** [comment]
Let's merge this and follow up in https://github.com/tendermint/tendermint/issues/3120

**Pull request merged**